InnoDB temporary files and temporary tables encryption

gl-sergei opened the pull request:
InnoDB is using temporary files for merge sort for online alter table.
It also creates temporary tables for online alter.

When destination table for online alter operation is encrypted merge
file will also be encrypted. InnoDB writing and reading merge files in
blocks. Each block encrypted with tablespace key of target tablespace.
Key version used to encrypt this block prepended to each block. Since
Percona Server does not support key rotation currently, key version is
always 0. Block encrypted with AES256 in CBC mode with IV consisting of
space_id+offset. IV is encrypted to make it non-predictable. Since CBC
can only encrypt data in multiples of block size, remainder of the
buffer is XOR'ed with encrypted IV.

When InnoDB needs to create temporary table for online alter operation
on encrypted table, temp table will be encrypted as well.

https://blueprints.launchpad.net/percona-server/+spec/innodb-temp-files

http://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1211/

percona-ysorokin commented on include/my_aes.h:
```diff
@@ -131,6 +131,46 @@ int my_aes_get_size(uint32 source_length, enum my_aes_opmode mode);
+  @param key_length     [in]  Length of the key. Will handle keys of any length
+  @param mode           [in]  encryption mode
+  @param iv             [in]  16 bytes initialization vector.
+  @param padding        [in]  if padding needed.
```
no such parameter

percona-ysorokin commented on include/my_aes.h:
```diff
@@ -131,6 +131,46 @@ int my_aes_get_size(uint32 source_length, enum my_aes_opmode mode);
+  @param key_length     Length of the key. Will handle keys of any length
+  @param mode           encryption mode
+  @param iv             16 bytes initialization vector.
+  @param padding        if padding needed.
```
no such parameter

percona-ysorokin commented on mysql-test/t/innodb_onlinealter_encryption-master.opt:
```diff
@@ -0,0 +1 @@
+$KEYRING_PLUGIN_OPT $KEYRING_PLUGIN_LOAD --keyring_file_data=$MYSQL_TMP_DIR/mysecret_keyring
```
no eol

percona-ysorokin commented on mysys_ssl/my_aes.cc:
```diff
@@ -57,3 +57,97 @@ void my_aes_create_key(const unsigned char *key, uint key_length,
+                         enum my_aes_opmode mode, const unsigned char *iv)
+{
+  unsigned char eiv[MY_AES_BLOCK_SIZE];
+  uint32 num_blocks= (source_length / MY_AES_BLOCK_SIZE) * MY_AES_BLOCK_SIZE;
```
misleading name, I would suggest something like `aligned_source_len`

percona-ysorokin commented on mysys_ssl/my_aes.cc:
```diff
@@ -57,3 +57,97 @@ void my_aes_create_key(const unsigned char *key, uint key_length,
+    unsigned char mask[MY_AES_BLOCK_SIZE];
+    uint32 len= source_length - num_blocks;
+
+    DBUG_ASSERT(dst_len == static_cast<int>(num_blocks));
```
should be outside of `if` block?

gl-sergei replied:
Outside of the block it can be false actually. `DBUG_ASSERT(dst_len == static_cast<int>(source_length));` should be true outside of the `if` block :)

gl-sergei replied:
Hmm... I got and idea.. Yes, it can be placed *before* `if` the block.

percona-ysorokin commented on mysys_ssl/my_aes.cc:
```diff
@@ -57,3 +57,97 @@ void my_aes_create_key(const unsigned char *key, uint key_length,
   }
 }
+
+int my_aes_encrypt_nopad(const unsigned char *source, uint32 source_length,
```
Can we implement this by using one of the AES block modes which do not require input message to be a multiple of AES_BLOCK_SIZE?
CFB, OFB or CTR for instance.

gl-sergei replied:
`my_aes.h` only provides ECB and CBC for yassl. We build PS with OpenSSL, but I think we still need to be compatible. InnoDB page encryption using CBC. I would prefer CTR of course. MariaDB provides choice between CTR and CBC (CTR only available for file_key_management plugin though) with CBC being default. Trick with last partial block is taken from MariaDB.

laurynas-biveinis replied:
It's OK to provide degraded functionality / missing features for YaSSL builds if that makes the OpenSSL implementation better.

robgolebiowski replied:
Guys, Mode of operations have nothing to do with block sizes. AES will always demand input to be multiple of 128 bits. If it is not multiple the padding is needed. Mode of operations (apart from ECB) are just the ways of getting random encryption. ECB is the only mode where no randomness is introduced into the output, thus each time when I encrypt text A with key K I will get B. ECB can be only used when text A is random - for instance INNODB is using it to encrypt innodb tablespace keys - and this is fine as keys are random. About which mode we should be using, 
there is a bit of discussion on this topic, but I think we should chose CBC. The reason being that in case of CTR we need to be sure to be always encrypting each input block with unique counter, the counter cannot repeat. In past gurus like Bruce Schneier (https://books.google.pl/books?id=nnqdQsrZnJgC&pg=PT141&lpg=PT141&dq=Bruce+Schneier+cbc+or+ctr&source=bl&ots=6GXUxllCq7&sig=ca_lQu9fkiu0Kf_qcBhhYkXVF2Q&hl=en&sa=X&ved=0ahUKEwiK-9Hy5c_XAhVKCpoKHXCZDOIQ6AEIXTAJ#v=onepage&q=Bruce%20Schneier%20cbc%20or%20ctr&f=false) would recommend CTR, but now they go with CBC. It safer this way with CTR it is quite possible one will repeat the counter.

gl-sergei replied:
Streaming modes such as CTR or OFB do not require input to be multiple of block size. They actually perform AES operations on key and XOR plaintext with produced stream (somewhat simplified explanation :) ). So the resulting length is always the same as input length.

robgolebiowski replied:
Yeah, my bad. I forgot that also CTR and OFB can be used as stream ciphers (however they still use block cipher inside).

percona-ysorokin commented on mysys_ssl/my_aes.cc:
```diff
@@ -57,3 +57,97 @@ void my_aes_create_key(const unsigned char *key, uint key_length,
+    if (dst_len == MY_AES_BAD_DATA)
+      return dst_len;
+
+    for (uint32 i= 0; i < len; i++)
```
Is this "xoring with encrypted eiv" a well-known approach or a clever workaround? If yes, please give a reference.

gl-sergei replied:
It actually very similar to how CTR mode works. It takes counter, encrypts it and XOR with plaintext, which allows to parallelize encryption.

percona-ysorokin commented on mysys_ssl/my_aes.cc:
```diff
@@ -57,3 +57,97 @@ void my_aes_create_key(const unsigned char *key, uint key_length,
+                         enum my_aes_opmode mode, const unsigned char *iv)
+{
+  unsigned char eiv[MY_AES_BLOCK_SIZE];
+  uint32 num_blocks= (source_length / MY_AES_BLOCK_SIZE) * MY_AES_BLOCK_SIZE;
```
as in `_encrypt_`: misleading name

gl-sergei commented:
Fixed comment, added newline, renamed variable and moved DBUG_ASSERT few lines upward.

laurynas-biveinis commented on mysql-test/t/innodb_onlinealter_encryption.test:
```diff
@@ -0,0 +1,74 @@
+-- source include/have_innodb.inc
+-- source include/not_valgrind.inc
```
Why?

gl-sergei replied:
cleaned up

laurynas-biveinis commented on mysql-test/t/innodb_onlinealter_encryption.test:
```diff
@@ -0,0 +1,74 @@
+-- source include/have_innodb.inc
+-- source include/not_valgrind.inc
+-- source include/not_embedded.inc
+-- source include/not_windows.inc
```
Why?

gl-sergei replied:
cleaned up

laurynas-biveinis commented on mysql-test/t/innodb_onlinealter_encryption.test:
```diff
@@ -0,0 +1,74 @@
+CREATE TABLE t4 (id INT, a VARCHAR(255)) engine=InnoDB;
+CREATE TABLE t5 (id INT NOT NULL PRIMARY KEY, a TEXT(500), b VARCHAR(255), FULLTEXT(b)) ENGINE=InnoDB ENCRYPTION='y';
+CREATE TABLE t6 (id INT, a TEXT(500), b VARCHAR(255), FULLTEXT(b)) ENGINE=InnoDB;
+CREATE TABLE t7 (id INT NOT NULL PRIMARY KEY, a VARCHAR(255)) ENGINE=InnoDB row_format=compressed ENCRYPTION='y';
```
Consistent upper case for row_format clause

laurynas-biveinis commented on mysql-test/t/innodb_onlinealter_encryption.test:
```diff
@@ -0,0 +1,74 @@
+CREATE TABLE t1 (id INT NOT NULL PRIMARY KEY, a VARCHAR(255)) ENGINE=InnoDB ENCRYPTION='y';
+CREATE TABLE t2 (id INT NOT NULL PRIMARY KEY, a VARCHAR(255)) ENGINE=InnoDB;
+CREATE TABLE t3 (id INT, a VARCHAR(255)) ENGINE=InnoDB ENCRYPTION='y';
+CREATE TABLE t4 (id INT, a VARCHAR(255)) engine=InnoDB;
```
Consistent upper case engine=

laurynas-biveinis commented on mysql-test/t/innodb_onlinealter_encryption.test:
```diff
@@ -0,0 +1,74 @@
+commit;
+set autocommit=1;
+
+ALTER TABLE t1 ADD COLUMN b int default 2;
```
Here and everywhere consistent uppercase (ideally the innodb_insert_proc definition and call above too)

laurynas-biveinis commented on mysql-test/t/innodb_onlinealter_encryption.test:
```diff
@@ -0,0 +1,74 @@
+-- source include/have_innodb.inc
```
IIRC there is now some have_keyring_*.inc too, please check and source

gl-sergei replied:
There is a `include/have_keyring_vault_plugin.inc`, do you suggest to add `have_keyring_file.inc` as well?

laurynas-biveinis replied:
It looks like this testcase needs have_keyring_file.inc and not have_keyring_vault_plugin.inc ?

gl-sergei replied:
I mean we do have `have_keyring_vault_plugin.inc` in our test suite, but not `have_keyring_file.inc` yet. I will create it.

laurynas-biveinis commented on mysql-test/t/innodb_onlinealter_encryption.test:
```diff
@@ -0,0 +1,74 @@
+SHOW CREATE TABLE t7;
+
+--echo # Restarting server
+let $restart_parameters = restart: $KEYRING_PLUGIN_OPT $KEYRING_PLUGIN_LOAD --keyring_file_data=$MYSQL_TMP_DIR/mysecret_keyring
```
The arguments seem to match the -master.opt file exactly. Are they still required to be specified here in that case?

laurynas-biveinis commented on storage/innobase/handler/handler0alter.cc:
```diff
@@ -5569,6 +5570,31 @@ ha_innobase::prepare_inplace_alter_table(
+	/* create_table_info_t::innobase_table_flags does not set encryption
+	flags. There are places where it is done afterwards, there are places
+	where it isn't done. We need to inspect all code paths and check if
+	encryption flag cen be set in one place. */
```
"can"

laurynas-biveinis commented on storage/innobase/row/row0ftsort.cc:
```diff
@@ -195,6 +196,7 @@ row_fts_psort_info_init(
 	fts_psort_t*		merge_info = NULL;
 	ulint			block_size;
 	ibool			ret = TRUE;
+	bool			encrypted = false;
```
Declare at initialization, as "const"?

laurynas-biveinis commented on storage/innobase/row/row0ftsort.cc:
```diff
@@ -267,6 +271,31 @@ row_fts_psort_info_init(
+				psort_info[j].crypt_alloc[i] =
+					static_cast<row_merge_block_t*>(
+						ut_malloc_nokey(
+							block_size + 1024));
```
IIRC we might need 4K for some O_DIRECT setups, please check - but if this is a copy-paste from a different temp file allocation doing same I/O, then OK to leave as-is

gl-sergei replied:
yes. it is copy-paste from
``` c++
			/* Need to align memory for O_DIRECT write */
			psort_info[j].block_alloc[i] =
				static_cast<row_merge_block_t*>(ut_malloc_nokey(
					block_size + 1024));
```

laurynas-biveinis commented on storage/innobase/row/row0ftsort.cc:
```diff
@@ -318,6 +347,10 @@ row_fts_psort_info_destroy(
 				ut_free(psort_info[j].merge_file[i]);
+
+				if (psort_info[j].crypt_alloc[i]) {
+					ut_free(psort_info[j].crypt_alloc[i]);
```
Does ut_free DTRT with NULL arg, allowing to remove "if" here?

gl-sergei replied:
I think ut_free is OK with NULL args. Changed. Not sure about `dealloc_large`, so I left that as it was

laurynas-biveinis commented on storage/innobase/row/row0ftsort.cc:
```diff
@@ -733,6 +766,7 @@ fts_parallel_tokenization(
 	ibool			processed = FALSE;
 	merge_file_t**		merge_file;
 	row_merge_block_t**	block;
+	row_merge_block_t**	crypt_block;
```
Declare at initialization?

laurynas-biveinis commented on storage/innobase/row/row0ftsort.cc:
```diff
@@ -1517,6 +1561,7 @@ row_fts_merge_insert(
 	mrec_buf_t**		buf;
 	int*			fd;
 	byte**			block;
+	byte**			crypt_block;
```
Declare at initialization?

laurynas-biveinis commented on storage/innobase/row/row0merge.cc:
```diff
@@ -259,12 +261,115 @@ class index_tuple_info_t {
+						write */
+	ulint			space_id,	/*!< in: tablespace id */
+	const byte*		input_buf,	/*!< in: input buffer */
+	byte*			crypted_buf)	/*!< out: crypted buffer */
```
"encrypted"

laurynas-biveinis commented on storage/innobase/row/row0merge.cc:
```diff
@@ -259,12 +261,115 @@ class index_tuple_info_t {
+{
+	const ulint key_version = 0;
+	int dstlen;
+	os_offset_t ofs = (os_offset_t)srv_sort_buf_size * (os_offset_t)offset;
```
What is the allowed maximum offset value here? Is it possible to overflow os_offset_t?

laurynas-biveinis commented on storage/innobase/row/row0merge.cc:
```diff
@@ -259,12 +261,115 @@ class index_tuple_info_t {
+	os_offset_t ofs = (os_offset_t)srv_sort_buf_size * (os_offset_t)offset;
+
+	/* Store key_version at the begining of the input buffer. InnoDB
+	does not support tablespace key rotation for now. */
```
This is not a "tablespace" is it?

laurynas-biveinis commented on storage/innobase/row/row0merge.cc:
```diff
@@ -259,12 +261,115 @@ class index_tuple_info_t {
+		iv);
+
+	if (dstlen == MY_AES_BAD_DATA) {
+		ib::fatal()
```
fatal? Maybe only the ALTER TABLE should be failed?

gl-sergei replied:
Replaced fatal with error, not sure how to cover it with a test case though.

laurynas-biveinis commented on storage/innobase/row/row0merge.cc:
```diff
@@ -259,12 +261,115 @@ class index_tuple_info_t {
+						write */
+	ulint			space_id,	/*!< in: tablespace id */
+	const byte*		input_buf,	/*!< in: input buffer */
+	byte*			crypted_buf)	/*!< out: crypted buffer */
```
"decrypted"

laurynas-biveinis commented on storage/innobase/row/row0merge.cc:
```diff
@@ -259,12 +261,115 @@ class index_tuple_info_t {
+{
+	ulint key_version;
+	int dstlen;
+	os_offset_t ofs = (os_offset_t)srv_sort_buf_size * (os_offset_t)offset;
```
What is the allowed maximum offset value here? Is it possible to overflow os_offset_t?

gl-sergei replied:
os_offset_t should be 64bit. But it is probably OK to overflow it here since it is just a part of IV.

laurynas-biveinis commented on storage/innobase/row/row0merge.cc:
```diff
@@ -259,12 +261,115 @@ class index_tuple_info_t {
+	key_version = mach_read_from_4((byte *)input_buf);
+
+	/* InnoDB does not support tablespace key rotation for now */
+	ut_a(key_version == 0);
```
Can we just fail the ALTER TABLE instead?

laurynas-biveinis commented on storage/innobase/row/row0merge.cc:
```diff
@@ -259,12 +261,115 @@ class index_tuple_info_t {
+		iv);
+
+	if (dstlen == MY_AES_BAD_DATA) {
+		ib::fatal()
```
Same comment re. fatal as above

laurynas-biveinis commented on storage/innobase/row/row0merge.cc:
```diff
@@ -259,12 +261,115 @@ class index_tuple_info_t {
+			<< ". Can't continue!";
+	}
+
+	return true;
```
It only returns true - either errors should be returned if possible to implement above review comments, either make it void. The caller only handles true return correctly.

laurynas-biveinis commented on storage/innobase/row/row0merge.cc:
```diff
@@ -1172,6 +1307,10 @@ row_merge_read_rec(
 	DBUG_ENTER("row_merge_read_rec");
 
+	if (b == &block[0]) {
+		b+= ROW_MERGE_RESERVE_SIZE;
```
Space before +=

laurynas-biveinis commented on storage/innobase/row/row0merge.cc:
```diff
@@ -1377,6 +1521,10 @@ row_merge_write_rec(
 		+ rec_offs_data_size(offsets);
 
+	if (b == &block[0]) {
+		b+= ROW_MERGE_RESERVE_SIZE;
```
Space before +=

laurynas-biveinis commented on storage/innobase/row/row0merge.cc:
```diff
@@ -1434,6 +1590,10 @@ row_merge_write_eof(
 		    fd, ulong(*foffs)));
 
+	if (b == &block[0]) {
+		b+= ROW_MERGE_RESERVE_SIZE;
```
Space before +=

laurynas-biveinis commented on storage/innobase/row/row0merge.cc:
```diff
@@ -4336,7 +4546,9 @@ row_merge_build_indexes(
 {
 	merge_file_t*		merge_files;
 	row_merge_block_t*	block;
+	row_merge_block_t*	crypt_block;
```
Declare at initialization

laurynas-biveinis commented on storage/innobase/row/row0merge.cc:
```diff
@@ -4369,6 +4581,21 @@ row_merge_build_indexes(
+	if (dict_table_is_encrypted(new_table)) {
+		crypt_block = static_cast<row_merge_block_t*>(
+			alloc.allocate_large(
+				3 * srv_sort_buf_size, &crypt_pfx));
```
This might be ridiculously defensive but please consider (the upstream might rewrite the same and we wouldn't notice): extract srv_sort_buf_size sysvar max value to a define, add a static assert that 3 * max does not overflow os_offset_t

gl-sergei replied:
Currently `sort_buffer_size` max value is 64 MiB, so it is very unlike to overflow. But if you insist, I will add static assert

laurynas-biveinis replied:
I don't insist :)

laurynas-biveinis commented on storage/innobase/row/row0merge.cc:
```diff
@@ -4462,6 +4689,10 @@ row_merge_build_indexes(
 	sorting and inserting. */
 
+	DBUG_EXECUTE_IF(
+		"ib_merge_wait_after_read",
```
Please don't :) Luckily, this does not seem to have any associated testcase (something not imported from MariaDB?). Can it be removed, or if a testcase import is missed, replaced with something else?

gl-sergei replied:
Test cases were cleaned up and use existing upstream DBUG_SYNC points

laurynas-biveinis commented on storage/innobase/row/row0merge.cc:
```diff
@@ -4545,9 +4776,13 @@ row_merge_build_indexes(
 				sort_idx, table, col_map, 0};
 
+			DBUG_EXECUTE_IF(
+				"ib_merge_wait_after_sort",
```
Likewise

robgolebiowski commented on mysys_ssl/my_aes.cc:
```diff
@@ -57,3 +57,99 @@ void my_aes_create_key(const unsigned char *key, uint key_length,
+  if (key_length == 64)
+    ecb_mode= my_aes_256_ecb;
+  else if (key_length == 32)
+    ecb_mode= my_aes_128_ecb;
```
This seems to be wrong. In AES 128 and 256 mean key sizes. Thus key length of 64 bytes means aes 512 and key length of 32 bytes means aes 256.

gl-sergei replied:
Good catch. Fixed

robgolebiowski commented on mysys_ssl/my_aes.cc:
```diff
@@ -57,3 +57,99 @@ void my_aes_create_key(const unsigned char *key, uint key_length,
+  enum my_aes_opmode ecb_mode;
+  int dst_len;
+
+  if (key_length == 64)
```
This seems to be wrong. In AES 128 and 256 mean key sizes. Thus key length of 64 bytes means aes 512 and key length of 32 bytes means aes 256.

gl-sergei commented:
Last week MariaDB 10.2.29 has been released with support for online alter logs encryption. Merge file encryption has changed significantly as well. I will review that changes and update this PR. This PR should be put on hold for now.

gl-sergei commented:
I pushed some changes. Implementation slightly changed. There is no more key version stored with each block. Instead single encryption key is used. See https://jira.mariadb.org/browse/MDEV-12634. Added variable `innodb_encrypt_online_alter_logs`. Online alter logs are encrypted as well.

http://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1302/

gl-sergei commented:
For AES modes, lets discuss our options. I suggest to use CTR when available otherwise (YaSSL?) revert to CBC.

robgolebiowski commented on include/my_aes.h:
```diff
@@ -131,6 +131,44 @@ int my_aes_get_size(uint32 source_length, enum my_aes_opmode mode);
+  @return              size of encrypted data, or negative in case of error
+*/
+
+int my_aes_encrypt_nopad(const unsigned char *source, uint32 source_length,
```
In binlog_encryption PR I am introducing the same functionality. The difference is only that I am introducing almost the same thing as MariaDB as I need the "context" feature they introduced. So I will be pushing this into repo and this functionalities will overlap. With what I am pushing the same will be possible by setting NO_PAD flag. However, I think the refactoring will needed to be done by one who pushes last

robgolebiowski commented on mysql-test/suite/innodb/t/innodb_onlinealter_encryption-master.opt:
```diff
@@ -0,0 +1,4 @@
+$KEYRING_PLUGIN_OPT
```
Tests that require keyring plugins shoud be extracted into include files. There should be a test in innodb testsuite that loads keyring_file and sources the test as include file and there should be a test in keyring_vault  that loads keyring_vault plugin and sources the include file. For exmple see table_encypt_2 test case.

laurynas-biveinis commented on mysql-test/suite/innodb/r/innodb_onlinealter_encryption.result:
```diff
@@ -0,0 +1,127 @@
+CREATE TABLE t1 (id INT NOT NULL PRIMARY KEY, a VARCHAR(255)) ENGINE=InnoDB ENCRYPTION='y';
```
online_alter in the testcase name

laurynas-biveinis commented on mysql-test/suite/innodb/r/innodb_rowlog.result:
```diff
@@ -0,0 +1,41 @@
+CREATE TABLE t1(id INT NOT NULL PRIMARY KEY AUTO_INCREMENT, credit_card VARCHAR(200), private VARCHAR(50)) ENGINE=InnoDB ENCRYPTION='y';
```
row_log in the testcase name

laurynas-biveinis commented on mysql-test/suite/innodb/t/innodb_rowlog.test:
```diff
@@ -0,0 +1,71 @@
+# Verify that both merge blocks and row log blocks are encryted and decrypted
+#
+
+SELECT variable_value > 0 FROM information_schema.global_status WHERE variable_name = 'innodb_num_pages_encrypted';
```
Can you query PFS instead? (add have_perfschema.inc too then)

laurynas-biveinis commented on mysql-test/suite/innodb/t/innodb_rowlog.test:
```diff
@@ -0,0 +1,71 @@
+#
+# Force key rotation and create second index for same table
+#
+connect (con2,localhost,root,,);
```
--source include/count_sessions.inc

laurynas-biveinis commented on mysql-test/suite/innodb/t/innodb_rowlog.test:
```diff
@@ -0,0 +1,71 @@
+DROP TABLE t2;
+
+connection default;
+disconnect con2;
```
--source include/wait_until_count_sessions.inc somewhere below this line

laurynas-biveinis commented on mysys_ssl/my_aes.cc:
```diff
@@ -57,3 +57,103 @@ void my_aes_create_key(const unsigned char *key, uint key_length,
+  else
+    return MY_AES_BAD_DATA;
+
+  my_aes_encrypt(iv, MY_AES_BLOCK_SIZE, eiv,
```
Check return value?

laurynas-biveinis commented on mysys_ssl/my_aes.cc:
```diff
@@ -57,3 +57,103 @@ void my_aes_create_key(const unsigned char *key, uint key_length,
+
+  DBUG_ASSERT(aligned_src_len <= source_length);
+
+  my_aes_encrypt(iv, MY_AES_BLOCK_SIZE, eiv,
```
Likewise?

laurynas-biveinis commented on storage/innobase/row/row0log.cc:
```diff
@@ -205,6 +215,138 @@ struct row_log_t {
+	/** Encryption algorithm */
+	Encryption::Type	encryption_type;
+
+	/** Encrypt key */
```
Encryption key

laurynas-biveinis commented on storage/innobase/row/row0log.cc:
```diff
@@ -205,6 +215,138 @@ struct row_log_t {
+	/** Encrypt key */
+	byte			encryption_key[ENCRYPTION_KEY_LEN];
+
+	/** Encrypt key length*/
```
Encryption key length

laurynas-biveinis commented on storage/innobase/srv/srv0srv.cc:
```diff
@@ -2754,6 +2767,9 @@ DECLARE_THREAD(srv_master_thread)(
 		}
+
+		/* Enable row log encryption if it is set */
+		log_tmp_enable_encryption_if_set();
```
The corresponding plugin variable is not dynamic - shouldn't this be one-time initialisation thing?

laurynas-biveinis commented on storage/innobase/row/row0log.cc:
```diff
@@ -205,6 +215,138 @@ struct row_log_t {
+
+crypt_info_t crypt_info = crypt_info_t();
+
+/** Find out is temporary log files encrypted.
```
Find out if temporary log files are encrypted

laurynas-biveinis commented on storage/innobase/row/row0log.cc:
```diff
@@ -205,6 +215,138 @@ struct row_log_t {
+struct crypt_info_t {
+
+	/** Encryption algorithm */
+	Encryption::Type	encryption_type;
```
Essentially duplicates each other with srv_encrypt_online_alter_logs. Not sure whether it needs fixing, and how

gl-sergei replied:
Right now we have implemented only AES, but maybe in future there will be more algorithms to choose from?

laurynas-biveinis commented on storage/innobase/row/row0log.cc:
```diff
@@ -205,6 +215,138 @@ struct row_log_t {
+}
+
+/** Check the row log encryption is enabled or not.
+It will try to enable the row log encryption. */
```
It will enable, there is no "try" :)

laurynas-biveinis commented on storage/innobase/row/row0merge.cc:
```diff
@@ -4370,6 +4471,21 @@ row_merge_build_indexes(
 		DBUG_RETURN(DB_OUT_OF_MEMORY);
 	}
 
+	/* Get crypt data from tablespace if present. */
```
Comment not clear, there is no "tablespace" here?

laurynas-biveinis commented on storage/innobase/row/row0merge.cc:
```diff
@@ -4370,6 +4471,21 @@ row_merge_build_indexes(
+	/* Get crypt data from tablespace if present. */
+	row_merge_block_t* crypt_block = NULL;
+
+	/* If tablespace is encrypted, allocate additional buffer for
```
Likewise

gl-sergei commented:
Rebased on 5.7 HEAD and addressed comments. https://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1385/

Once binlog encryption get merged, I will update tests as suggested

laurynas-biveinis commented:
@gl-sergei, binlog encryption has been merged. Does this PR need updating?

gl-sergei commented:
rebased on latest 5.7, updated test cases to run with keyring_file and keyring_vault, switched to use of my_crypt instead of my_aes

https://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1595/

laurynas-biveinis commented on mysql-test/include/innodb_row_log_encryption.inc:
```diff
@@ -0,0 +1,82 @@
+--source include/have_debug_sync.inc
+--source include/have_perfschema.inc
+
+--source include/count_sessions.inc
```
Not before the server restart

laurynas-biveinis commented on mysql-test/include/innodb_row_log_encryption.inc:
```diff
@@ -0,0 +1,82 @@
+#
+connect (con2,localhost,root,,);
+SET DEBUG_SYNC= 'now WAIT_FOR opened';
+# TODO: once we add key rotation and scrubbing make sure to add it to this test
```
Track this in JIRA on the main feature tickets

laurynas-biveinis replied:
Have you logged something in JIRA?

laurynas-biveinis commented on mysql-test/suite/perfschema/r/show_sanity.result:
```diff
@@ -672,6 +674,10 @@ where (show_mode = "JUNK: GLOBAL-ONLY")
 and source = "I_S.SESSION_STATUS")
 order by show_mode, source, variable_name;
 SHOW_MODE	SOURCE	VARIABLE_NAME
+5.7	P_S.SESSION_STATUS	Innodb_encryption_n_merge_blocks_decrypted
```
Was show_sanity updated correctly? I.e. no need to mention the new vars in the testcase itself anywhere?

laurynas-biveinis commented on storage/innobase/include/row0log.h:
```diff
@@ -255,6 +256,44 @@ row_log_estimate_work(
 	const dict_index_t*	index);
 #endif /* HAVE_PSI_STAGE_INTERFACE */
 
+bool
```
Header comment

laurynas-biveinis replied:
MY_ATTRIBUTE((warn_unused_result))

laurynas-biveinis commented on storage/innobase/include/row0log.h:
```diff
@@ -255,6 +256,44 @@ row_log_estimate_work(
+bool
+log_tmp_is_encrypted();
+
+void
```
Header comment

laurynas-biveinis commented on storage/innobase/include/row0log.h:
```diff
@@ -255,6 +256,44 @@ row_log_estimate_work(
+@param[in]	offs		offset to block
+@param[in]	space_id	tablespace id
+@return whether the operation succeeded */
+bool
```
MY_ATTRIBUTE((warn_unused_result)) (and this will show one unchecked call site)

laurynas-biveinis commented on storage/innobase/include/row0log.h:
```diff
@@ -255,6 +256,44 @@ row_log_estimate_work(
+@param[in]	offs		offset to block
+@param[in]	space_id	tablespace id
+@return whether the operation succeeded */
+bool
```
MY_ATTRIBUTE((warn_unused_result))

laurynas-biveinis commented on storage/innobase/include/row0merge.h:
```diff
@@ -362,12 +365,14 @@ row_merge_file_create(
+@param[in]	dup		descriptor of index being created
+@param[in,out]	file		file containing index entries
+@param[in,out]	block		3 buffers
+@param[in,out]	crypt_block	crypted file buffer
```
"encrypted". I am not sure what's up with MariaDB inventing verb "to crypt"

laurynas-biveinis commented on storage/innobase/log/log0log.cc:
```diff
@@ -56,6 +56,7 @@ Created 12/9/1995 Heikki Tuuri
 #include "trx0roll.h"
 #include "srv0mon.h"
 #include "sync0sync.h"
+#include "row0log.h"
```
Sorry if I asked this before, but why this? There are new declarations in row0log.h but none appear to be moved from somewhere else so that otherwise untouched source file starts needing them

laurynas-biveinis commented on storage/innobase/row/row0log.cc:
```diff
@@ -205,6 +215,158 @@ struct row_log_t {
+
+	if (res != MY_AES_OK) {
+		ib::error()
+			<< "Unable to encrypt data-block "
```
No hyphen

laurynas-biveinis commented on storage/innobase/row/row0log.cc:
```diff
@@ -205,6 +215,158 @@ struct row_log_t {
+
+	if (res != MY_AES_OK) {
+		ib::error()
+			<< "Unable to decrypt data-block "
```
Likewise

laurynas-biveinis commented:
Please also file a feature request on JIRA and reference its id in the commit message header

gl-sergei commented:
Rebased and fixed:
- count sessions in the test case
- header comments
- grammar errors
- removed row0log.h (probably leftover from previous implementation)
- added new status variables into corresponding tables instead of re-recordling show_sanoty.result

https://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1607/

laurynas-biveinis commented on storage/innobase/row/row0log.cc:
```diff
@@ -179,6 +182,7 @@ struct row_log_t {
 	dict_table_t*	table;	/*!< table that is being rebuilt,
 				or NULL when this is a secondary
 				index that is being created online */
+	dict_index_t*	index;	/*!< index to be build */
```
s/build/built

laurynas-biveinis commented on storage/innobase/row/row0log.cc:
```diff
@@ -3202,10 +3417,23 @@ row_log_allocate(
+		log->crypt_head = static_cast<byte *>(os_mem_alloc_large(&size));
+		log->crypt_tail = static_cast<byte *>(os_mem_alloc_large(&size));
+
+		if (!log->crypt_head || !log->crypt_tail) {
```
Doesn't log->crypt_head != NULL, log->crypt_tail == NULL result in a memory leak for the former?

gl-sergei replied:
`row_log_free` will deallocate both as I read

laurynas-biveinis commented on storage/innobase/row/row0merge.cc:
```diff
@@ -1637,6 +1690,7 @@ NULL if old_table == new_table
 ULINT_UNDEFINED if none is added
 @param[in,out]	sequence	autoinc sequence
 @param[in,out]	block		file buffer
+@param[in,out]	crypt_block	crypted file buffer
```
encrypted

laurynas-biveinis commented on storage/innobase/row/row0merge.cc:
```diff
@@ -2708,13 +2776,15 @@ row_merge_read_clustered_index(
+@param[in]	dup		descriptor of index being created
+@param[in]	file		file containing index entries
+@param[in,out]	block		3 buffers
+@param[in,out]	crypt_block	crypted file buffer
```
encrypted

laurynas-biveinis commented on storage/innobase/row/row0merge.cc:
```diff
@@ -2809,17 +2885,21 @@ row_merge_blocks(
+@param[in]	index		index being created
+@param[in]	file		input file
+@param[in,out]	block		3 buffers
+@param[in,out]	crypt_block	crypted file buffer
```
encrypted

laurynas-biveinis commented on storage/innobase/row/row0merge.cc:
```diff
@@ -2894,6 +2980,8 @@ row_merge_blocks_copy(
 @param[in]	dup		descriptor of index being created
 @param[in,out]	file		file containing index entries
 @param[in,out]	block		3 buffers
+@param[in,out]	crypt_block	crypted file buffer
```
encrypted

laurynas-biveinis commented:
...and/or @robgolebiowski review.

Also please file a JIRA issue for this, reference in the commit message header

gl-sergei commented:
repushed with grammar fixes in comments

gl-sergei commented:
Tracked here: https://jira.percona.com/browse/PS-3819

gl-sergei commented:
I created both, just in the wrong project, then moved to PS ... here is the second one  https://jira.percona.com/browse/PS-3820

percona-ysorokin commented on mysql-test/suite/perfschema/r/show_sanity.result:
```diff
@@ -412,6 +412,7 @@ SHOW_MODE	SOURCE	VARIABLE_NAME
 5.6	I_S.SESSION_VARIABLES	INNODB_COMPRESSED_COLUMNS_THRESHOLD
 5.6	I_S.SESSION_VARIABLES	INNODB_COMPRESSED_COLUMNS_ZIP_LEVEL
 5.6	I_S.SESSION_VARIABLES	INNODB_DEADLOCK_DETECT
+5.6	I_S.SESSION_VARIABLES	INNODB_ENCRYPT_ONLINE_ALTER_LOGS
```
Put `INNODB_ENCRYPT_ONLINE_ALTER_LOGS` as an exception to `show_sdanity.test`

gl-sergei replied:
done

percona-ysorokin commented on storage/innobase/include/row0log.h:
```diff
@@ -255,6 +256,51 @@ row_log_estimate_work(
+log_tmp_enable_encryption_if_set();
+
+/** Encrypt a temporary file block.
+@param[in]	src		block to encrypt or decrypt
```
`block to encrypt`

gl-sergei replied:
fixed this comment and corresponding one in .cc file

percona-ysorokin commented on storage/innobase/include/row0log.h:
```diff
@@ -255,6 +256,51 @@ row_log_estimate_work(
+	ulint			space_id);
+
+/** Decrypt a temporary file block.
+@param[in]	src		block to encrypt or decrypt
```
`block to decrypt`

percona-ysorokin commented on storage/innobase/row/row0ftsort.cc:
```diff
@@ -267,6 +268,31 @@ row_fts_psort_info_init(
+							psort_info[j]
+							.crypt_alloc[i], 1024));
+
+				if (!psort_info[j].crypt_block[i]) {
```
Is it OK to `ut_align()` on `NULL`s? If not, may be move this check just after `ut_malloc_nokey()`?

gl-sergei replied:
NULL remains NULL after alignment because it is already aligned. I see no issues here.

gl-sergei replied:
hmm.. however .. .`ut_align` implementation seems to disagree with me. It contains `ut_ad(ptr);`

gl-sergei replied:
moved check before `ut_align`

gl-sergei replied:
BTW, upstream code just few lines above has the same problem. I did not touch it.

percona-ysorokin commented on storage/innobase/row/row0log.cc:
```diff
@@ -205,6 +215,158 @@ struct row_log_t {
+	mach_write_to_8(iv, space_id);
+	mach_write_to_8(iv + 8, offs);
+
+	int res = my_aes_crypt(MY_AES_ECB,
```
Does `my_aes_crypt()` allow to do inplace encryption?

gl-sergei replied:
Basically it is just a wrapper for `EVP_EncryptUpdate`, which does not allow `in` and `out` to partially overlap, but it does support in-place encryption. I am cheating here of course because I know that `iv` size equals to cipher block size and must be enough to accommodate the result, but I am doing it consciously.

percona-ysorokin commented on storage/innobase/row/row0log.cc:
```diff
@@ -205,6 +215,158 @@ struct row_log_t {
+
+	int res = my_aes_crypt(MY_AES_ECB,
+		ENCRYPTION_FLAG_ENCRYPT | ENCRYPTION_FLAG_NOPAD,
+		iv, MY_AES_BLOCK_SIZE, iv, &dst_len,
```
Does `dst_len` need to be initialized with max out buffer size?

gl-sergei replied:
`dst_len` is output only. Here and below as well.

percona-ysorokin commented on storage/innobase/row/row0log.cc:
```diff
@@ -205,6 +215,158 @@ struct row_log_t {
+
+	res = my_aes_crypt(MY_AES_CBC,
+		ENCRYPTION_FLAG_ENCRYPT | ENCRYPTION_FLAG_NOPAD,
+		src_block, size, dst_block, &dst_len,
```
The same question re `dst_len`.

percona-ysorokin commented on storage/innobase/row/row0log.cc:
```diff
@@ -205,6 +215,158 @@ struct row_log_t {
+
+	int res = my_aes_crypt(MY_AES_ECB,
+		ENCRYPTION_FLAG_ENCRYPT | ENCRYPTION_FLAG_NOPAD,
+		iv, MY_AES_BLOCK_SIZE, iv, &dst_len,
```
The same questions re inplace encryption and `dst_len`.

percona-ysorokin commented on storage/innobase/row/row0log.cc:
```diff
@@ -205,6 +215,158 @@ struct row_log_t {
+
+	res = my_aes_crypt(MY_AES_CBC,
+		ENCRYPTION_FLAG_DECRYPT | ENCRYPTION_FLAG_NOPAD,
+		src_block, size, dst_block, &dst_len,
```
The same question re `dst_len`.

gl-sergei commented:
All comments addressed, Jenkins: https://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1624/
